### DISCUSSION TITLE
Problem with block-diagonalizing the Bosonic Hubbard model using user-defined basis

### DISCUSSION THREAD

**Discussion by P2C2E1993** [QuSpin/QuSpin#636] [Q&A]
Hello,

I began using Qspin recently, and have been trying to block-diagonalize the Bosonic Hubbard model with a user-defined basis. I am trying it with a 10 site, 10 particle system. I was able to obtain the hamiltonian in matrix form, but the result I got from trying to block-diagonalize it doesn't seem correct.

I am attaching my code, along with the outputs as a reference.

L = 10 # length of chain
N = L # number of sites
Nb = L # number of bosons
J = 1.0# Hubbard hopping strength
U = 1.0 # Hubbard interaction

s = np.arange(L)
T = s%L #Translation along the lattice
P = L-T-1 #Reflection operation

#basis = boson_basis_1d(L,Nb) # full boson basis
basis_symmetric = boson_basis_general(L, Nb, Tblock = (T,0), Pblock= (P,0)) # user-defined basis that obeys the symmetries of H

# define site-coupling lists
int_list_1 = [[mu-0.5*U,i] for i in range(N)] # interaction $-U/2 \sum_i n_i$
int_list_2 = [[0.5*U,i,i] for i in range(N)] # interaction: $U/2 \num_i n_i^2$

# setting up hopping lists
hop_list = [[-J, i, (i+1)%N] for i in range(0 , N, 1)]
hop_list_hc = [[J.conjugate(),i,j] for J,i,j in hop_list] # add h.c. terms

# set up static and dynamic lists
static = [
            ["+-",hop_list], # hopping
            ["-+",hop_list_hc], # hopping h.c.
            ["nn",int_list_2], # U n_i^2
            ["n",int_list_1] # -U n_i
         ]
dynamic = [] # no dynamic operators

#H=hamiltonian(static, dynamic, basis=basis, dtype=np.complex128, check_symm=False)
H=hamiltonian(static, dynamic, basis=basis_symmetric, dtype=np.complex128, check_symm=False)

H_sparse = H.tocsr(time=0)
plt.spy(H_sparse, marker=',')
print(H_sparse.shape, H_sparse.nnz)

#Outputs
![BH_Hamil_L10N10](https://github.com/QuSpin/QuSpin/assets/123269457/d4d16d57-bb9c-49c1-a8b2-94ef58b95a2d)
![BH_Hamil_L10N10_symmetry_reduced](https://github.com/QuSpin/QuSpin/assets/123269457/8b10ad14-9811-487d-bf01-4090031a1b23)

I made use of the example provided in the boson_basis_general documentation, but the output for the symmetry-reduced matrix doesn't look like it has any blocks. I would like to know if I have specified the symmetries correctly. The Bose-Hubbard model has cyclic permutation symmetry and reflection symmetry, similar to an n-sided regular polygon (Dihedral group).

**Comment by mgbukov** [comment]
Hi, when u specify a symmetry block in the _basis_ object, the Hamiltonian
is constructed *only* within this block; to get a block-diagonal matrix,
you'd have to construct every sector independently, and then put them
manually in the blocks of a larger matrix; this is to save memory.

On Thu, Jul 27, 2023 at 1:59 PM P2C2E1993 ***@***.***> wrote:

> Hello,
>
> I began using Qspin recently, and have been trying to block-diagonalize
> the Bosonic Hubbard model with a user-defined basis. I am trying it with a
> 10 site, 10 particle system. I was able to obtain the hamiltonian in matrix
> form, but the result I got from trying to block-diagonalize it doesn't seem
> correct.
>
> I am attaching my code, along with the outputs as a reference.
>
> L = 10 # length of chain
> N = L # number of sites
> Nb = L # number of bosons
> J = 1.0# Hubbard hopping strength
> U = 1.0 # Hubbard interaction
>
> s = np.arange(L)
> T = s%L #Translation along the lattice
> P = L-T-1 #Reflection operation
>
> #basis = boson_basis_1d(L,Nb) # full boson basis
> basis_symmetric = boson_basis_general(L, Nb, Tblock = (T,0), Pblock=
> (P,0)) # user-defined basis that obeys the symmetries of H
> [image: BH_Hamil_L10N10]
> <https://user-images.githubusercontent.com/123269457/256528718-d4d16d57-bb9c-49c1-a8b2-94ef58b95a2d.png>
> [image: BH_Hamil_L10N10_symmetry_reduced]
> <https://user-images.githubusercontent.com/123269457/256528723-8b10ad14-9811-487d-bf01-4090031a1b23.png>
> define site-coupling lists
>
> int_list_1 = [[mu-0.5
> *U,i] for i in range(N)] # interaction $-U/2 \sum_i n_i$ int_list_2 =
> [[0.5*U,i,i] for i in range(N)] # interaction: $U/2 \num_i n_i^2$
> setting up hopping lists
>
> hop_list = [[-J, i, (i+1)%N] for i in range(0 , N, 1)]
> hop_list_hc = [[J.conjugate(),i,j] for J,i,j in hop_list] # add h.c. terms
> set up static and dynamic lists
>
> static = [
> ["+-",hop_list], # hopping
> ["-+",hop_list_hc], # hopping h.c.
> ["nn",int_list_2], # U n_i^2
> ["n",int_list_1] # -U n_i
> ]
> dynamic = [] # no dynamic operators
>
> #H=hamiltonian(static, dynamic, basis=basis, dtype=np.complex128,
> check_symm=False)
> H=hamiltonian(static, dynamic, basis=basis_symmetric, dtype=np.complex128,
> check_symm=False)
>
> H_sparse = H.tocsr(time=0)
> plt.spy(H_sparse, marker=',')
> print(H_sparse.shape, H_sparse.nnz)
>
> I made use of the example provided in the boson_basis_general
> documentation, but the output for the symmetry-reduced matrix doesn't look
> like it has any blocks. I would like to know if I have specified the
> symmetries correctly. The Bose-Hubbard model has cyclic permutation
> symmetry and reflection symmetry, similar to an n-sided regular polygon
> (Dihedral group).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/QuSpin/QuSpin/discussions/636>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADHSUBWZSDRN5LFBFDSVYN3XSJJ3HANCNFSM6AAAAAA2Z66BPI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Marin Bukov, PhD
Research Group Leader
Max Planck Institute for the Physics of Complex Systems
Noethnitzer Str. 38
Dresden 01187, Germany

url: https://www.pks.mpg.de/nqd
<http://quantum-dynamics.phys.uni-sofia.bg/>
